core/shared/src/main/scala/fs2/util/Catenable.scala

```diff
@@ -5,70 +5,134 @@ import Catenable._
+
+  /** Folds over the elements from left to right using the supplied initial value and function. */
+  final def foldLeft[B](z: B)(f: (B, A) => B): B = {
+    var result = z
+    foreach(a => result = f(result, a))
+    result
+  }
+
+  /** Applies the supplied function to each element, left to right. */
+  final def foreach(f: A => Unit): Unit = {
+    var c: Catenable[A] = this
+    var rights: List[Catenable[A]] = Nil
+    while (c ne null) {
+      c match {
+        case Empty =>
+          rights match {
+            case Nil => c = null
+            case h :: t => c = h; rights = t
+          }
+        case Single(a) =>
+          f(a)
+          c = if (rights.isEmpty) Empty else rights.reverse.reduceLeft((x, y) => Append(y,x))
+          rights = Nil
+        case Append(l, r) => c = l; rights = r :: rights
+      }
+    }
   }
 
-  def map[B](f: A => B): Catenable[B] = Catenable.fromSeq(toStream.map(f))
+  /** Converts to a list. */
+  final def toList: List[A] = {
+    val builder = List.newBuilder[A]
+    foreach { a => builder += a; () }
+    builder.result
+  }
+
+  override def toString = "Catenable(..)"
 }
 
 object Catenable {
-  private case object Empty extends Catenable[Nothing]
-  private final case class Single[A](a: A) extends Catenable[A]
-  private final case class Append[A](left: Catenable[A], right: Catenable[A]) extends Catenable[A]
+  private final case object Empty extends Catenable[Nothing] {
+    def isEmpty: Boolean = true
+  }
+  private final case class Single[A](a: A) extends Catenable[A] {
+    def isEmpty: Boolean = false
+  }
+  private final case class Append[A](left: Catenable[A], right: Catenable[A]) extends Catenable[A] {
+    def isEmpty: Boolean = false // b/c `append` constructor doesn't allow either branch to be empty
+  }
 
+  /** Empty catenable. */
   val empty: Catenable[Nothing] = Empty
 
+  /** Creates a catenable of 1 element. */
   def single[A](a: A): Catenable[A] = Single(a)
 
-  def append[A](c: Catenable[A], c2: Catenable[A]): Catenable[A] = c match {
-    case Empty => c2
-    case _ => c2 match {
-      case Empty => c
-      case _ => Append(c,c2)
-    }
-  }
+  /** Appends two catenables. */
+  def append[A](c: Catenable[A], c2: Catenable[A]): Catenable[A] =
+    if (c.isEmpty) c2
+    else if (c2.isEmpty) c
+    else Append(c, c2)
 
+  /** Creates a catenable from the specified sequence. */
   def fromSeq[A](s: Seq[A]): Catenable[A] =
     if (s.isEmpty) empty
-    else s.view.map(single).reduceRight(Append(_,_))
+    else s.view.reverse.map(single).reduceLeft((x, y) => Append(y, x))
 
-  def apply[A](as: A*): Catenable[A] = fromSeq(as)
+  /** Creates a catenable from the specified elements. */
+  def apply[A](as: A*): Catenable[A] = {
+    // Assumption: `as` is small enough that calling size doesn't outweigh benefit of calling empty/single
+    as.size match {
+      case 0 => empty
+      case 1 => single(as.head)
+      case n if n <= 1024 =>
```

Comment:
Yeah, `scala.collection.mutable.WrappedArray`. Maybe 1024 is too high -- we could limit to 16 or 32 and cover 99.9% of cases.